frontend/packages/polkadot-wallet/components/ConnectWalletModal.tsx

```diff
@@ -0,0 +1,310 @@
+                name={network.name}
+                icon={network.icon}
+                onClick={(networkId: NetworkId) => {
+                    networkId === NetworkId.Ethereum ? setWalletConnectStep(WalletConnectStep.SelectEthereumWallet) :
+                        setWalletConnectStep(WalletConnectStep.SelectedDotsamaWallet)
+                }}
+            />
+        ))
+    }, [networks]);
+
+    const polkadotWalletsList = useCallback(() => {
+        return supportedPolkadotWallets.map(wallet => (
+            <ConnectListItem
+                onClick={(walletId: SupportedWalletId) => {
+                    onConnectPolkadotWallet(walletId).then((walletConnected) => {
+                        setWalletConnectStep(WalletConnectStep.SelectDotsamaAccount);
+                    });
+                }}
+                name={wallet.name}
+                icon={wallet.icon}
+                id={wallet.walletId}
+            />
+        ))
+    }, [supportedPolkadotWallets, onConnectPolkadotWallet]);
+
+    const ethereumWalletsList = useCallback(() => {
+        return supportedEthereumWallets.map(wallet => (
+            <ConnectListItem
+                onClick={(walletId: ConnectorType) => {
+                    onConnectEthereumWallet(walletId);
+                }}
+                name={wallet.name}
+                icon={wallet.icon}
+                id={wallet.walletId}
+            />
+        ))
+    }, [supportedEthereumWallets, onConnectEthereumWallet])
+
+    const title = useMemo(() => {
+        switch (walletConnectStep) {
+            case WalletConnectStep.SelectNetwork:
+                return "Wallets";
+            case WalletConnectStep.SelectedDotsamaWallet:
+                return "Connect Dotsama";
+            case WalletConnectStep.SelectEthereumWallet:
+                return "Connect EVM";
+            case WalletConnectStep.SelectDotsamaAccount:
+                return "Select Account";
+        }
+    }, [walletConnectStep]);
+
+    const description = useMemo(() => {
+        switch (walletConnectStep) {
+            case WalletConnectStep.SelectNetwork:
+                return "Select a network to continue";
+            case WalletConnectStep.SelectedDotsamaWallet:
+            case WalletConnectStep.SelectEthereumWallet:
+            case WalletConnectStep.SelectDotsamaAccount:
+                return "Select a wallet to connect with";
+        }
+    }, [walletConnectStep]);
+
+    const takeOneStepBack = useCallback(() => {
+        switch (walletConnectStep) {
+            case WalletConnectStep.SelectDotsamaAccount:
+                setWalletConnectStep(WalletConnectStep.SelectedDotsamaWallet);
+                break;
+            case WalletConnectStep.SelectedDotsamaWallet:
+            case WalletConnectStep.SelectEthereumWallet:
+                setWalletConnectStep(WalletConnectStep.SelectNetwork);
+                break;
+        }
+    }, [walletConnectStep]);
```

Comment:
The computation is not heavy to rely on useMemo, all of these can be moved to their own pure functions and to be used here.